File: vercors/src/main/java/vct/col/ast/RecordType.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import java.util.*;

public class RecordType extends Type {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  private String name[];
  private Type type[];
  public RecordType(String name[],Type type[]){
    this.name=Arrays.copyOf(name,name.length);
    this.type=Arrays.copyOf(type,type.length);
  }

  public int getFieldCount(){ return name.length; }
  
  public String getFieldName(int i){ return name[i]; }

  public Type getFieldType(int i){return type[i]; }
  
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
      }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 
  @Override
  public boolean supertypeof(ProgramUnit context, Type t) {
    // TODO Auto-generated method stub
    return false;
  }
  @Override
  public <T> T accept_simple(TypeMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }

}

File: vercors/src/main/java/vct/col/ast/RecordType.scala
package vct.col.ast

import vct.col.util.VisitorHelper

/** A single type entry of a record type (only used in `RecordType`). */
case class RecordTypeEntry(val recordName:String, val recordType:Type)

/**
 * AST node that represents record types, that is, pairs of variable names (identifiers)
 * and their types. 
 */
case class RecordType(val types:List[RecordTypeEntry]) extends Type with VisitorHelper {
  /** Instantiates a new record type from separate lists of `names` and `types`. */
  def this(names:List[String], types:List[Type]) = 
    this((names, types).zipped map (new RecordTypeEntry(_,_)))
  
  /** @return The number of fields in this record. */
  def fieldCount = types.length
  
  /** @return The name of the {@code i}-th record field. */
  def fieldName(i:Int) = types.apply(i).recordName
  
  /** @return The type of the {@code i}-th record field. */
  def fieldType(i:Int) = types.apply(i).recordType
  
  /** @return Always `false`: no type can extend (inherit from) a record type. */
  override def supertypeof(context:ProgramUnit, t:Type) = false
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def accept_simple[T](m:TypeMapping[T]) = handle_standard(() => m.map(this))
}
